bin/supervise.sh

```diff
@@ -0,0 +1,100 @@
+                               # ms part of fractional seconds w/o leading 0
+                               ms: . | modf | .[0] | "\(.)" | .[1:5]
+                           } |
+                           "\(.datetime)\(.ms)Z",
+                status: $status,
+                message: .
+            }'
+    else
+        # provide a _very_ minimal fallback here
+        read message
+        echo "{\"status\":\"$level\",\"message\":\"$message\"}"
+    fi
+}
+
+# https://github.com/beacon-biosignals/K8sDeputy.jl/blob/b62e1858a4083ffc8f9f7b10fcb60a77896ae13e/src/graceful_termination.jl#L14
+IPC_DIR="${DEPUTY_IPC_DIR:-/run}"
+
+# https://github.com/beacon-biosignals/K8sDeputy.jl/blob/b62e1858a4083ffc8f9f7b10fcb60a77896ae13e/src/graceful_termination.jl#L21-L29
+get_pid()
+{
+    PID_FILE="${DEPUTY_IPC_DIR}/julia-entrypoint.pid"
+    # TODO: timeout.  not likely to be critical since we don't call this unitl termination
+    # is requested
+    # until [ -f "${PID_FILE}" ]; do
+    #     sleep 0.1
+    # done
+    if [[ ! -f $PID_FILE ]]; then
+        echo "Failed to find PID file at $PID_FILE" | logger error
+        exit 1
+    fi
+    echo "reading PID from $PID_FILE" | logger debug
+    read -r SUPERVISED_PID <"${DEPUTY_IPC_DIR}/julia-entrypoint.pid"
+    if [[ ! $SUPERVISED_PID =~ ^[0-9]+$ ]]; then
+        echo "PID file $PID_FILE does not contain a numeric PID: $SUPERVISED_PID" | logger error
+        exit 1
+    fi
+    echo "supervised process has PID $SUPERVISED_PID" | logger debug
+}
+
+# https://github.com/beacon-biosignals/K8sDeputy.jl/blob/b62e1858a4083ffc8f9f7b10fcb60a77896ae13e/src/graceful_termination.jl#L16-L19
+get_socket()
+{
+    SOCKET_PATH="${DEPUTY_IPC_DIR}/graceful-terminator.${SUPERVISED_PID}.socket"
+    # TODO: timeout.  not likely to be critical since we don't call this unitl termination
+    # is requested
+    # until [ -e "$SOCKET_PATH" ]; do
+    #     sleep 0.1
+    # done
+    if [[ ! -S $SOCKET_PATH ]]; then
+        echo "Expected socket at $SOCKET_PATH; got something else. $SUPERVISED_PID may be a zombie now" | logger error
+        exit 1
+    fi
+    echo "using socket at $SOCKET_PATH" | logger debug
+}
+
+terminate_supervised()
+{
+    echo "TERM trapped, stopping" | logger
+    # we parse these at termination time because they may not be ready at startup, and
+    # because this matches the behavior of `K8sDeputy.graceful_terminate`
+    get_pid
+    get_socket
+    # https://github.com/beacon-biosignals/K8sDeputy.jl/blob/b62e1858a4083ffc8f9f7b10fcb60a77896ae13e/src/graceful_termination.jl#L143-L144
+    nc -U "$SOCKET_PATH" <<<"terminate"
+    wait $child
+    exit $?
+}
+
+JQ=$(command -v jq)
+if [[ -z $JQ ]]; then
+    echo "logging works best with jq" | logger warn
+fi
+
+if ! command -v nc >/dev/null; then
+    echo "supervise.sh requires netcat (nc)" | logger error
+    exit 1
+fi
+
+# start background process
+"$@" &
+child=$!
+
+trap "terminate_supervised" TERM
+
+wait $child
```

Comment:
The `get_pid` function can be eliminated if we update `terminate_supervised` to take the child PID as an argument:
```suggestion
child_pid=$!

trap "terminate_supervised $child_pid" TERM

wait $child_pid
```
The reason for having the `julia-entrypoint.pid` at was just for `K8sDeputy.graceful_terminate` which would start as a separate process with no context as to which PID to gracefully terminate. As this script knows the PID we can bypass that step entirely.

Reply:
the tricky thing is that if you're runnign julia via `juliaup`, then `$child` is not the PID that's used in the socket file name.

Reply:
and more generally if we're running anything _other_ than "just" `julia` via this supervision entrypoint, we may end up in a similar situation

Reply:
Very good points. Definitely worth adding a comment about this.